aws/rust-runtime/aws-auth/src/provider/lazy_caching.rs

```diff
@@ -0,0 +1,463 @@
+        pub fn default_credential_expiration(mut self, duration: Duration) -> Self {
+            self.default_credential_expiration = Some(duration);
+            self
+        }
+
+        /// Creates the [`LazyCachingCredentialsProvider`].
+        pub fn build(self) -> LazyCachingCredentialsProvider {
+            let default_credential_expiration = self
+                .default_credential_expiration
+                .unwrap_or(DEFAULT_CREDENTIAL_EXPIRATION);
+            assert!(
+                default_credential_expiration >= DEFAULT_CREDENTIAL_EXPIRATION,
+                "default_credential_expiration must be at least 15 minutes"
+            );
+            LazyCachingCredentialsProvider::new(
+                self.refresh.expect("refresh provider is required"),
+                self.refresh_timeout.unwrap_or(DEFAULT_REFRESH_TIMEOUT),
+                default_credential_expiration,
+            )
+        }
+    }
+}
+
+// Allows us to abstract time for tests.
+trait TimeProvider: Clone + Send + Sync + 'static {
+    fn now(&self) -> SystemTime;
+}
+
+#[derive(Copy, Clone)]
+struct SystemTimeProvider;
+
+impl TimeProvider for SystemTimeProvider {
+    fn now(&self) -> SystemTime {
+        SystemTime::now()
+    }
+}
+
+/// Returns whether or not the `credentials` are expired based on the given time.
+/// This will panic if the credentials don't have an `expiry` set.
+fn expired(credentials: &Credentials, now: SystemTime) -> bool {
+    let expiration = credentials
+        .expiry()
+        .expect("refresh sets expiry if not given");
+    now > expiration
+}
+
+#[derive(Clone)]
+struct Inner<T: TimeProvider> {
+    time: T,
+    cache: Cache,
+    refresh: Arc<dyn AsyncProvideCredentials>,
+    refresh_timeout: Duration,
+    default_credential_expiration: Duration,
+}
+
+impl<T: TimeProvider> Inner<T> {
+    async fn refresh(&self) -> CredentialsResult {
+        let time = self.time.clone();
+        let default_credential_expiration = self.default_credential_expiration;
+        let future = self.refresh.provide_credentials();
+        self.cache
+            .refresh(|| async move {
+                let credentials = future.await?;
+                // If the credentials don't have an expiration time, then create a default one
+                let credentials = if credentials.expiry().is_none() {
+                    Credentials::new(
+                        credentials.access_key_id(),
+                        credentials.secret_access_key(),
+                        credentials.session_token().map(|s| s.to_string()),
+                        Some(time.now() + default_credential_expiration),
+                        "lazy_caching_default_credential_expiration",
+                    )
+                } else {
+                    credentials
+                };
+                Ok(credentials)
+            })
+            .await
+    }
+
+    async fn needs_refresh(&self, now: SystemTime) -> bool {
```

Comment:
`needs_refresh` is a fairly confusing name for a function that will actually _clear_ the state if a refresh is required

Reply:
I would probably combine `needs_refresh` and `refresh` (or call `needs_refresh` from inside refresh`). I don't think having them as separate is necessarily good.

Reply:
This is gone now.